File: leetcode_java/src/main/java/com/programapprentice/app/GasStation.java
package com.programapprentice.app;

/**
 * Created with IntelliJ IDEA.
 * User: ruijiang
 * Date: 8/8/14
 * Time: 5:34 PM
 */
/**
 * There are N gas stations along a circular route, where the amount of gas at station i is gas[i].
 *
 * You have a car with an unlimited gas tank and it costs cost[i] of gas to travel from station i to
 * its next station (i+1). You begin the journey with an empty tank at one of the gas stations.
 *
 * Return the starting gas station's index if you can travel around the circuit once, otherwise return -1.
 *
 * Note:
 * The solution is guaranteed to be unique.
 */
public class GasStation {

    // Time complexity: O(n^2)
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int n = gas.length;
        int cur = 0;
        int i, j;
        for(i = 0; i < n; i++) {
            cur = 0;
            for(j = 0; j < n; j++) {
                cur = cur + gas[(i + j) % n] - cost[(i + j) % n];
                if(cur < 0) {
                    break;
                }
            }
            if(j == n) {
                return i;
            }
        }
        return -1;
    }
}

File: leetcode_cpp/src/gas_station.cc
/*
 * gas_station.cc
 *
 *  Created on: Dec 22, 2014
 *      Author: ruijiang
 */

#include "gas_station.h"
#include <iostream>

using namespace::std;

int Solution::canCompleteCircuit(vector<int> &gas, vector<int> &cost) {
    int length = gas.size();
    int left = 0;
    int start = 0;
    int end = 0;
    while(start<length) {
        left = 0;
        for(end = 0; end < length; end++) {
            left += gas[(start + end) % length] - cost[(start + end) % length];

            if(left < 0) {
                break;
            }
        }
        cout << "left is " << left << " end is " << end << " start is " << start << endl;
        if(end == length && left >= 0) {
            return start;
        }
        start = start + end + 1;
    }
    return -1;
}
